torch/distributed/_shard/checkpoint/state_dict_saver.py

```diff
@@ -0,0 +1,136 @@
+            tensor_write_requests += st_write_reqs
+            metadata.state_dict_metadata[fqn] = st_md
+        elif isinstance(obj, Tensor):
+            write_reqs, tensor_md = _prepare_tensor_write(obj, fqn, storage_key_to_fqn)
+            if write_replicated_data:
+                tensor_write_requests += write_reqs
+            metadata.state_dict_metadata[fqn] = tensor_md
+        else:
+            bytes_io = io.BytesIO()
+            # This produces incomplete MD for rank > 0 since we won't populate bytes_io.
+            # This is ok since only rank == 0 uses this data
+            if write_replicated_data:
+                torch.save(obj, bytes_io)
+            byte_write_reqs, bytes_md = _prepare_bytes_write(bytes_io, fqn, storage_key_to_fqn)
+            if write_replicated_data:
+                bytes_write_requests += byte_write_reqs
+            metadata.state_dict_metadata[fqn] = bytes_md
+
+    storage_keys: Dict[str, int] = {
+        req.storage_key: req.tensor.nelement() * req.tensor.element_size()
+        for req in tensor_write_requests
+    }
+
+    return (metadata, storage_keys, bytes_write_requests, tensor_write_requests)
+
+
+def save_state_dict(
+    state_dict: Dict[str, Any],
+    storage_writer: StorageWriter,
+) -> None:
+    """
+    Save a distributed model in SPMD style.
+
+    This function is different from ``torch.save()`` as it handles
+    ``ShardedTensor`` by having each rank only save their local shards.
+
+    To produce a state_dict with ShardedTensor instances you must call
+    ``_register_state_dict_hook`` on the top module with value
+    `torch.distributed._shard.sharded_tensor.state_dict_hook` prior to
+    calling `state_dict()` on the top module.
+
+    There is no guarantees of Backwards Compatibility across PyTorch versions
+    for saved state_dicts.
+
+    Args:
+        state_dict (Dict[str, Any]) : A state_dict
+        storage_writer (StorageWriter): Instance of StorageWrite use to perform writes.
+
+    Example:
+        >>> my_model = MyModule()
+        >>> # We must call this function prior to state_dict()
+        >>> my_model._register_state_dict_hook(state_dict_hook)
+
+        >>> model_state_dict = my_model.state_dict()
+
+        >>> fs_storage_writer = torch.distributed._shard.checkpoint.FileSystemWriter("/checkpoint/1")
+        >>> torch.distributed._shard.checkpoint.save_state_dict(
+        >>>     state_dict=model_state_dict,
+        >>>     storage_writer=fs_stroage_writer,
+        >>> )
+    """
+    (
+        metadata,
+        storage_keys,
+        bytes_write_requests,
+        tensor_write_requests,
+    ) = _prepare(state_dict)
+    storage_writer.prepare_storage(storage_keys=storage_keys)
+    storage_writer.write_metadata(metadata=metadata)
+    bytes_futures = storage_writer.write_bytes(bytes_write_requests)
+    tensor_futures = storage_writer.write_tensors(tensor_write_requests)
+    bytes_futures.wait()
+    tensor_futures.wait()
```

Comment:
This order is inverted from the doc:
The API doc says:

```
    1) prepare_storage with a dict of storage_key -> size. This can be used to
        pre-allocate storage.
    2) write_bytes
    3) write_tensors.
    4) (if rank == 0) write_metadata().
```

This doesn't do that nor does it do the rank 0 check?
Also this seems to write metadata even if writing tensor fails, which seems ok but also a bit odd.

Reply:
I changed the order to reflect the docs and make sure that failures are detectable.